storagemarket/provider_loop.go

```diff
@@ -389,56 +415,39 @@ func (p *Provider) run() {
 				// to the database but don't execute the deal. The deal
 				// will be executed when the Storage Provider imports the
 				// deal data.
-				dh, err := p.mkAndInsertDealHandler(deal.DealUuid)
-				if err != nil {
-					sendErrorResp(&acceptError{error: err, isSevereError: true, reason: "server error: creating deal handler"})
-					continue
-				}
+				go p.processDeal(deal, dealReq.rsp)
+				continue
+			}
 
-				aerr := p.processOfflineDealProposal(dealReq.deal, dh)
+			if deal.IsOffline && dealReq.isImport {
+				// The Storage Provider is importing offline deal data, so tag
+				// funds for the deal and execute it
+				aerr := p.processImportOfflineDealData(dealReq.deal)
 				if aerr != nil {
-					dh.close()
-					p.delDealHandler(deal.DealUuid)
 					sendErrorResp(aerr)
 					continue
 				}
 
-				// The deal proposal was successful. Send an Accept response to the client.
-				dealReq.rsp <- acceptDealResp{ri: &api.ProviderDealRejectionInfo{Accepted: true}}
-				// Don't execute the deal now, wait for data import.
-				continue
-			}
+				// set up deal handler so that clients can subscribe to deal update events
+				dh, err := p.mkAndInsertDealHandler(deal.DealUuid)
+				if err != nil {
+					sendErrorResp(&acceptError{error: err, isSevereError: true, reason: "server error: starting deal thread"})
+					continue
+				}
 
-			var aerr *acceptError
-			if deal.IsOffline {
-				// The Storage Provider is importing offline deal data, so tag
-				// funds for the deal and execute it
-				aerr = p.processImportOfflineDealData(dealReq.deal)
-			} else {
-				// Process a regular deal proposal
-				aerr = p.processDealProposal(dealReq.deal)
-			}
-			if aerr != nil {
-				sendErrorResp(aerr)
-				continue
-			}
+				// start executing the deal
+				_, err = p.startDealThread(dh, deal)
+				if err != nil {
+					sendErrorResp(&acceptError{error: err, isSevereError: true, reason: "server error: starting deal thread"})
+					continue
+				}
 
-			// set up deal handler so that clients can subscribe to deal update events
-			dh, err := p.mkAndInsertDealHandler(deal.DealUuid)
-			if err != nil {
-				sendErrorResp(&acceptError{error: err, isSevereError: true, reason: "server error: starting deal thread"})
-				continue
+				// send an accept response
+				dealReq.rsp <- acceptDealResp{&api.ProviderDealRejectionInfo{Accepted: true}, nil}
```

Comment:
Specify keys in struct and avoid setting zero value fields (i.e. `err`)?

```suggestion
				dealReq.rsp <- acceptDealResp{ri: &api.ProviderDealRejectionInfo{Accepted: true}}
```